src/transformers/generation/tf_logits_process.py

```diff
@@ -584,3 +584,109 @@ def _force_token(generation_idx):
+
+    Args:
+        generate_config (`GenerateConfig`):
+            The generate config used to generate the output. The following parameters are required:
+                eos_token_id (`int`, *optional*, defaults to 50257):
+                    The id of the *end-of-sequence* token.
+                no_timestamps_token_id (`int`, *optional*, defaults to 50363):
+                    The id of the `"<|notimestamps|>"` token.
+                max_initial_timestamp_index (`int`, *optional*, defaults to 1):
+                    Used to set the maximum value of the initial timestamp. This is used to prevent the model from
+                    predicting timestamps that are too far in the future.
+    """
+
+    def __init__(self, generate_config):  # support for the kwargs
+        self.eos_token_id = generate_config.eos_token_id
+        self.no_timestamps_token_id = generate_config.no_timestamps_token_id
+        self.timestamp_begin = generate_config.no_timestamps_token_id + 1
+
+        self.begin_index = len(generate_config.forced_decoder_ids) + 2
+        if generate_config.forced_decoder_ids[-1][1] == self.no_timestamps_token_id:
+            self.begin_index -= 1
+        self.max_initial_timestamp_index = generate_config.max_initial_timestamp_index
+
+    def __call__(self, input_ids, scores, cur_len):
+        def _un_force_token(current_token):
+            # scores[:,current_tokens ] = -float("inf")
+            batch_size = scores.shape[0]
+            new_scores = tf.identity(scores)
+            indices = tf.stack((tf.range(batch_size), tf.tile([current_token], [batch_size])), axis=1)
+            updates = tf.ones((batch_size,), dtype=scores.dtype) * -float("inf")
+            new_scores = tf.tensor_scatter_nd_update(new_scores, indices, updates)
+            return new_scores
+
+        def _force_tokens(current_tokens):
+            # scores[:, :] = -float("inf")
+            # scores[:,current_tokens ] = 0
+            batch_size = scores.shape[0]
+            new_scores = tf.ones_like(scores, dtype=scores.dtype) * -float("inf")
+            indices = tf.stack((tf.range(batch_size), tf.tile([current_tokens], [batch_size])), axis=1)
+            updates = tf.zeros((batch_size,), dtype=scores.dtype)
+            new_scores = tf.tensor_scatter_nd_update(new_scores, indices, updates)
+            return new_scores
+
+        # suppress <|notimestamps|> which is handled by without_timestamps
+        scores = _un_force_token(self.no_timestamps_token_id)  # scores[:, self.no_timestamps_token_id] = -float("inf")
+        if cur_len == self.begin_index - 1:
+            scores = _force_tokens(self.timestamp_begin)
+            return scores
+        # len(seq) == 1 corresponds to cur_len == self.begin_index + 2
+        last_was_timestamp = (cur_len >= self.begin_index + 2) & (input_ids[:, cur_len - 1] >= self.timestamp_begin)
+        penultimate_was_timestamp = (cur_len < self.begin_index + 3) | (
```

Comment:
Just for my understanding is the `cur_len < self.begin_index + 3` line because the first timestamp won't have a previous timestamp token prediction?

Reply:
Yes! I can add a comment for this 😉